README.md
# Установка npm-модулей
```bash
npm install
```

# Настройка проекта
В `gulpfile.js` в конфиг вынесены основные параметры:
```js
var CONFIG = {
	output: 'public', /* Корневая папка сайта */
	input: '.distr/', /* Корневая папка дистрибутива */
	pages: '.distr/pages', /* Структура сайта в дистрибутиве */
	templates: '.distr/templates', /* Шаблоны сайта в дистрибутиве */
	blocks: '.distr/blocks', /* Блоки сайта в дистрибутиве */
	proxyPortPhp: 8002, /* Прокси-порт для PHP-сервера */
	proxyPortBs: 8910, /* Прокси-порт для browserSync */
	useAutoprefixer: false, /* Autoprefixer по умолчанию выключен */
	reload: true /* Перезагрузка браузера по умолчанию влючена */
}
```

Для переопределения настроек можно использовать свой файл `gulpconfig.json` (в корне проекта), например, если какой-то порт уже занят или нескольких проектов должны быть запущены параллельно:
```js
{
	"proxyPortPhp": 8003,
	"proxyPortBs": 8911
}
```

# Сборка проекта
Для сборки проекта в `gulpfile.js` есть две основные задачи:

1. `default` - сборка для dev (выполянется по умолчанию), без сжатия, с живой перезагрузкой в браузере (browserSync, порт :8910).
```bash
npm run gulp
```

2. `build` - просто сборка, без запуска browserSync.
```bash
npm run gulp build
```
или со сжатием, для production (в случае использования докера)
```bash
NODE_ENV=production npm run gulp build
```

Также есть есть две дополнительные задачи:

1. `nosync` - сборка для dev без поднятия PHP-сервера и browserSync.
```bash
npm run gulp nosync
```

2. `noreload` - сборка без перезагрузки страницы (когда не нужно, чтобы после изменений фокус каждый раз прыгал к самому началу страницы).
```bash
npm run gulp noreload
```


## Структура проекта
`pages` - структура сайта (версии, подпапки, index.php, version.php и т.д.). Копируется 1-в-1 в папку `public`, с предварительным парсингом шаблонов.<br>
Здесь также находятся:<br>
- папки `css` с главными `*.scss`-файлами проекта с @import-ами стилей всех нужных блоков и шрифтами
- папки `js` с главными `*.js`-файлами проекта с require-ми скриптов всех нужных блоков (используется синтаксис плагина [gulp-include](https://www.npmjs.com/package/gulp-include#include-directives))

`templates` - шаблоны страниц с инклюдами блоков. В качестве шаблонизатора используется [Nunjucks](https://mozilla.github.io/nunjucks/templating.html) ([Примеры](https://css-tricks.com/killer-features-of-nunjucks/))

`blocks` - блоки шаблонов

Блок - это не обязательно HTML-блок. Можно заводить логические блоки, например, `common`, `icon`, `fonts` и т.д. Каждый блок и его подключаемые файлы должны находиться в отдельной папке.

В папке блока может находиться:
* `block.php`- [шаблон блока](http://prntscr.com/ki23p8)
* `macro.php`- [макросы блока](http://prntscr.com/ki22y8)
* `data.php`- [данные блока](http://prntscr.com/ki242n)
* `img` - папка для изображений блока (внутри возможна любая вложенность)
* `_style.scss` - основные стили блока.  В качестве шаблонизатора используется [SCSS](https://sass-scss.ru/documentation/rasshirenie_css/) ([Статья на русском с примерами](https://habr.com/post/96417/), [Песочница-компилятор](https://www.sassmeister.com), [Примеры с БЭМ](https://www.sassmeister.com/gist/be4b839c6d3ffb59a9868718cf81de55))
* `_responsive.scss` - адаптивные стили блока
* `script.js` - скрипты блока

Во всех php-файлах также используется Nunjucks.<br>
Внутри `blocks` и `img` возможна любая вложенность.<br>
Для групировки блоков для других страниц рекомендуется использовать префикс `@` с именем страницы, например, если в структуре есть папка `version1`, блоки можно сложить в папку `@version1`.


## Логика сборки
### HTML
1. Сборщик пропускает php-файлы в папке `.distr/pages` через шаблонизатор Nunjucks.
1. Когда встречается упоминание `extends`, включается обработка шаблона из папки `templates`.
1. В шаблоне при нахождении `include` парсится шаблон блока из указанной папки.
1. В папку `public` копируются скомпилированные php-файлы из папки `.distr/pages`.

### Стили
1. Сборщик пропускает scss-файлы из папок `.distr/pages/*/css` через шаблонизатор SCSS.
1. Файлы с подчёркиванием '_', указанные с помощью `@import`, инклюдятся и компилируются. Файлы без подчёркивания компилируются автоматически в алфавитном порядке.
1. В `public/*/css` копируются скомпилированные css-файлы.

### Изображения
* В папку `public/img/*` копируются файлы из папок `.distr/blocks/*/img` (изображения блока копируются в папку блока).

### Скрипты
1. Сборщик пропускает js-файлы из папок `.distr/pages/*/js` через плагин `gulp-include`.
1. В `public/*/js` копируются скомпилированные js-файлы.

### Другие файлы
* Все остальные файлы из папки `.distr/pages` копируются в `public` в соответствии со своей структурой.

CHANGELOG.md
# Изменения в сборочных файлах

## [27-02-2020]

### Добавлено
- Поддержка редиректа на зеркало

## [27-01-2020]

### Добавлено
- Поддержка в обработке редиректов разных комбинаций папки/подпапки - параметры/без параметров

## [27-12-2019]

### Добавлено
- Поддержка сжатия brotli
- При запросе jpg/jpeg/png файла, если рядом лежит webp с тем же именем, и если клиент поддерживает webp, то отдается webp. В проектах теперь достаточно оставлять только одну ссылку на картинки jpeg/jpg/png, а при компиляции все jpeg/jpg/png в webp копировать
- Добавлена параметр query в файле .helm/values.yaml при редиректе. Теперь можно сделать редирект на сайт с подстановкой переменных. Теперь можно сделать, например,  редирект с https://SynergyStartupForum.ru на https://SynergyStartupForum.com/?lang=ru

### Изменено
- В контейнере и в настройках php временная зона по умолчанию - Europe/Moscow
- В переменной REQUEST_URI теперь реальный путь и использовать кастомную переменную HTTP_X_ORIGINAL_REQUEST_URL не нужно
- Изменены настройки заголовков кеширования nginx
- Изменены заголовки X-UA-Compatible Referrer-Policy X-Content-Type-Options X-Frame-Options X-XSS-Protection Access-Control-Allow-Origin в nginx


.distr/templates/index.php
<?php 
$ROOT = $_SERVER['DOCUMENT_ROOT'] . '/';
include_once $ROOT . 'version.php';
?>

<!DOCTYPE html>
<html>

<head>
    <?php if (!empty($_SERVER['DOCUMENT_URI'])) : ?>
        <base href="//<?= $_SERVER['HTTP_HOST'] . str_replace(substr($_SERVER['DOCUMENT_URI'], 1, -9), '', $_SERVER['REQUEST_URI']) ?>">
    <?php else : ?>
        <base href="//<?= $_SERVER['HTTP_HOST'] ?>">
    <?php endif; ?>

    <meta charset="utf-8">
    <title><?= $title ?></title>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge">
    <meta name="SKYPE_TOOLBAR" content="SKYPE_TOOLBAR_PARSER_COMPATIBLE">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <meta property="og:title" content="<?= $title ?>">
    <meta property="og:description" content="">
    <meta property="og:url" content="">
    <meta property="og:image" content="">
    <link rel="" href="">

    <link href="img/common/vbc.png" type="image/x-icon" rel="icon">

    <link href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" rel="stylesheet">

    <link href="css/style.css" rel="stylesheet">
    <link href="css/responsive.css" rel="stylesheet">

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.14.0/css/all.css" integrity="sha384-HzLeBuhoNPvSl5KYnjx0BT+WB0QEEqLprO+NBkkk5gbc67FTaL7XIGa2w1L0Xbgc" crossorigin="anonymous">

</head>

<body>

    <div class="wrapper flex">
        {% block blocks %}
            {% include 'chat/block.php' %}
        {% endblock %}
    </div>

    <!-- Плагины и хелперы -->
    <script src="js/script.js"></script>

</body>

</html>

.distr/blocks/chat/block.php
<section class="chat">
	<div class="chat__container flex">
		<div class="chat__left">
			<div class="chat__left_header">
				<div class="chat__left_title">Бизнеc-чат</div>
				<div class="chat__left_close"><i class="fas fa-times"></i></div>
			</div>
			<div class="chat__left_subtitle">Продукты</div>
			<ul class="chat__left_list">
				<li class="chat__left_product">
					<div class="chat__left_product-top flex-between">
						<div class="chat__left_product-id">БГ(№345678)</div>
						<div class="chat__left_product-date">21.06 10:40</div>
					</div>
					<div class="chat__left_product-bottom flex-between">
						<div class="chat__left_product-name"><i class="fas fa-arrow-right"></i>Маркетплайс ВБЦ</div>
						<div class="chat__left_product-icons flex">
							<div class="chat__left_product-message flex">
								<div class="chat__left_product-img"><i class="fas fa-comment-alt"></i></div>
								<div class="chat__left_product-count bg-grey">3</div>
							</div>

							<div class="chat__left_product-info flex">
								<div class="chat__left_product-img"><i class="fas fa-file-alt"></i></div>
								<div class="chat__left_product-count bg-red">1</div>
							</div>
						</div>
					</div>
				</li>
			</ul>
		</div>
		<div class="chat__right">
			<div class="chat__right_content">
				<ul class="chat__right_list">
					<li class="chat__message chat__message_personal">
						<div class="chat__message_wrap">
							<div class="chat__message_header flex-between">
								<div class="chat__message_who">Маркетплейс ВБЦ / Менеджер / Агафьев Петр Николаевич</div>
								<div class="chat__message_time">10:40:23</div>
							</div>
							<div class="chat__message_text">По своей сути рыбатекст является альтернативой традиционному lorem ipsum, который вызывает у некоторых людей недоумение при попытках прочитать рыбу текст</div>
						</div>
					</li>
					<li class="chat__message chat__message_personal">
						<div class="chat__message_wrap">
							<div class="chat__message_header flex-between">
								<div class="chat__message_who">Маркетплейс ВБЦ / Менеджер / Агафьев Петр Николаевич</div>
								<div class="chat__message_time">10:40:23</div>
							</div>
							<div class="chat__message_text">По своей сути рыбатекст является альтернативой традиционному lorem ipsum, который вызывает у некоторых людей недоумение при попытках прочитать рыбу текст</div>
						</div>
					</li>
					<li class="chat__message chat__message_client">
						<div class="chat__message_wrap">
							<div class="chat__message_header flex-between">
								<div class="chat__message_who">Маркетплейс ВБЦ / Менеджер / Агафьев Петр Николаевич</div>
								<div class="chat__message_time">10:40:23</div>
							</div>
							<div class="chat__message_text">По своей сути рыбатекст является альтернативой традиционному lorem ipsum, который вызывает у некоторых людей недоумение при попытках прочитать рыбу текст</div>
						</div>
					</li>
					<li class="chat__message chat__message_personal">
						<div class="chat__message_wrap">
							<div class="chat__message_header flex-between">
								<div class="chat__message_who">Маркетплейс ВБЦ / Менеджер / Агафьев Петр Николаевич</div>
								<div class="chat__message_time">10:40:23</div>
							</div>
							<div class="chat__message_text">По своей сути рыбатекст является альтернативой традиционному lorem ipsum, который вызывает у некоторых людей недоумение при попытках прочитать рыбу текст</div>
						</div>
					</li>
					<li class="chat__message chat__message_personal">
						<div class="chat__message_wrap">
							<div class="chat__message_header flex-between">
								<div class="chat__message_who">Маркетплейс ВБЦ / Менеджер / Агафьев Петр Николаевич</div>
								<div class="chat__message_time">10:40:23</div>
							</div>
							<div class="chat__message_text">По своей сути рыбатекст является альтернативой традиционному lorem ipsum, который вызывает у некоторых людей недоумение при попытках прочитать рыбу текст</div>
						</div>
					</li>
					<div class="chat__newmessage">
						<div></div>
						<p>Новые сообщения</p>
						<div></div>
					</div>
					<li class="chat__message chat__message_client">
						<div class="chat__message_wrap">
							<div class="chat__message_header flex-between">
								<div class="chat__message_who">Маркетплейс ВБЦ / Менеджер / Агафьев Петр Николаевич</div>
								<div class="chat__message_time">10:40:23</div>
							</div>
							<div class="chat__message_upload">
								<div class="chat__message_upload_text">Загрузите такой-то документ с длинным предлинным названием</div>							
								<input type="file" id="fileInput" class="chat__message_upload_input">
								<label for="fileInput" class="chat__message_upload_file">
									<div>Загрузить</div>
								</label>
							</div>
						</div>
					</li>
				</ul>	
				<div class="chat__right_date">29 сентября</div>
			</div>

			<div class="chat__right_form">
				<form action="" class="chat__form">
					<div class="chat__input">
						<textarea name="textarea" class="chat__textarea" rows="5" placeholder="Введите текст сообщения ..."></textarea>
					</div>

					<div class="chat__row flex-between">
						<ul class="chat__actions">
							<li class="chat__actions_item">
								<a href="" class="chat__actions_link"><i class="fas fa-paperclip"></i>Загрузить файл</a>
							</li>
							<li class="chat__actions_item">
								<a href="" class="chat__actions_link"><i class="fas fa-file-medical"></i>Создать запрос</a>
							</li>
							<li class="chat__actions_item">
								<a href="" class="chat__actions_link"><i class="fas fa-file-signature"></i>Запустить подписание договора</a>
							</li>
						</ul>

						<div class="chat__button">
							<button type="button">Отправить <i class="fas fa-arrow-right"></i></button>
						</div>
					</div>
				</form>
			</div>

		</div>
	</div>
</section>

gulpfile.js
'use strict';

/* Плагины */
const gulp = require('gulp');
const gulpLoadPlugins = require('gulp-load-plugins');
const plugins = gulpLoadPlugins();
const browserSync = require('browser-sync').create();
const runSequence = require('run-sequence');
const del = require('del');
const combine = require('stream-combiner2').obj;
const os = require('os');
const fs = require('fs');


/* Переменные окружения */
const isDevelopment = !process.env.NODE_ENV || process.env.NODE_ENV != 'production';


/* Конфиг */
var CONFIG = {
	output: 'public', /* Корневая папка сайта */
	input: '.distr/', /* Корневая папка дистрибутива */
	pages: '.distr/pages', /* Структура сайта в дистрибутиве */
	templates: '.distr/templates', /* Шаблоны сайта в дистрибутиве */
	blocks: '.distr/blocks', /* Блоки сайта в дистрибутиве */
	proxyPortPhp: 8002, /* Прокси-порт для PHP-сервера */
	proxyPortBs: 8910, /* Прокси-порт для browserSync */
	useAutoprefixer: false, /* Autoprefixer по умолчанию выключен */
	reload: true /* Перезагрузка браузера по умолчанию влючена */
};

/* Если есть gulpconfig.json, берём параметры из него */
function loadConfig() {
	if (fs.existsSync(__dirname + '/./gulpconfig.json')) {
		let user_config = {};

		user_config = require('./gulpconfig.json');
		CONFIG = Object.assign({}, CONFIG, user_config);
		return CONFIG;
	}
}
loadConfig();


/* Задачи */

/* Задача для поднятия PHP-сервера */
gulp.task('connect', function (callback) {
	console.log('* Запуск gulp-connect-php *');

	let serverConfig = {
		base: CONFIG.output,
		port: CONFIG.proxyPortPhp
	};

	if ( os.type() == 'Windows_NT' ) {
		/* Для запуска под Windows должен быть установлен локальный PHP и путь к нему указан в %PATH% */
		serverConfig.bin = 'php.exe';
		serverConfig.ini = 'php.ini';
	}

	plugins.connectPhp.server(serverConfig, callback);
});


/* Задача для запуска browserSync, используя в качестве прокси PHP-сервер на порту 8001 */
gulp.task('browserSync', ['connect'], function() {
	console.log('* Запуск browserSync *');

	browserSync.init({
		proxy: '127.0.0.1:' + CONFIG.proxyPortPhp,
		open: false,
		reloadOnRestart: true,
		injectChanges: true,
		port: CONFIG.proxyPortBs
	});
});


/* Задача для очистки конечной сборки (удаляется всё в CONFIG.output) */
gulp.task('clean', function () {
	console.log('* Удаление предыдущей сборки *');

	return del(['**', '!' + CONFIG.input + '**'], {force: true, cwd: CONFIG.output});
});


/* Задача для картинок */
gulp.task('images', function () {
	console.log('* Копирование картинок *');

	return gulp.src(['**/img/**/*.*', '!**/img/**/*.ps*'], {cwd: CONFIG.blocks})
	.pipe( plugins.if( !isDevelopment, plugins.rev() ) )
	.pipe( plugins.rename(function (path) {
		let slash = '/';
		if ( os.type() == 'Windows_NT' ) slash = '\\';
		path.dirname = path.dirname.replace(slash + 'img', ''); /* Замена пути к картинкам для конечного пути: block/img/*.* -> img/block/*.* */
	}) )
	.pipe( gulp.dest(CONFIG.output + '/img/') )
	.pipe( plugins.if( !isDevelopment, combine( plugins.rev.manifest('images.json'), gulp.dest('manifest'), plugins.debug({title: 'manifest -> images.json'}) ) ) )
	;
});


/* Задача для JS */
gulp.task('scripts', function () {
	console.log('* Копирование скриптов *');

	return gulp.src('**/*.js', {cwd: CONFIG.pages})
	.pipe( plugins.include({
		includePaths: [CONFIG.blocks]
	}) )
	.pipe( plugins.if( !isDevelopment, plugins.rev() ) )
	.pipe( gulp.dest(CONFIG.output) )
	.pipe( plugins.if( !isDevelopment, combine( plugins.rev.manifest('js.json'), gulp.dest('manifest'), plugins.debug({title: 'manifest -> js.json'}) ) ) )
	;
});


/* Задача для CSS */
gulp.task('styles:css', function () {
	console.log('* Копирование стилей *');

	return gulp.src('**/*.scss', {cwd: CONFIG.pages})
	.pipe( plugins.if( isDevelopment, plugins.sourcemaps.init() ) )
	.pipe( plugins.sass({
		includePaths: [CONFIG.blocks],
		indentType: 'tab',
		indentWidth: 1,
		outputStyle: 'compact',
		outputStyle: !isDevelopment ? 'compressed' : 'expanded'
	}) )
	.pipe( plugins.if( CONFIG.useAutoprefixer, plugins.autoprefixer({
		browsers: ['last 10 versions']
	}) ) )
	.pipe( plugins.if( isDevelopment, plugins.sourcemaps.write() ) )
	.pipe( plugins.if( !isDevelopment, plugins.rev() ) )
	.pipe( gulp.dest(CONFIG.output) )
	.pipe( plugins.if( !isDevelopment, combine( plugins.rev.manifest('css.json'), gulp.dest('manifest'), plugins.debug({title: 'manifest -> css.json'}) ) ) )
	;
});


/* Задача для рендеринга шаблонов Nunjucks */
gulp.task('nunjucks', ['styles:css', 'scripts'], function() {
	console.log('* Рендеринг шаблонов (Nunjucks) *');

	return gulp.src('**/*.php', {cwd: CONFIG.pages})
	.pipe( plugins.nunjucksRender({
		path: [CONFIG.templates, CONFIG.blocks],
		inheritExtension: true,
		throwOnUndefined: true
	})
	)
	.pipe( gulp.dest(CONFIG.output) )
	;
});


/* Задача для копирования остальных файлов */
gulp.task('other.pages', ['nunjucks'], function() {
	console.log('* Копирование остальных файлов *');

	return gulp.src(['**/**', '!**/*.{php,scss}'], {cwd: CONFIG.pages})
	.pipe( gulp.dest(CONFIG.output) )
	;
});


/* Задача для замены имён файлов в HTML и CSS */
gulp.task('revreplace', ['nunjucks'], function(callback) {
	if ( isDevelopment ) return callback;

	console.log('* Замена имён файлов *');

	let
	manifestCss = gulp.src('manifest/css.json'),
	manifestImages = gulp.src('manifest/images.json'),
	manifestJs = gulp.src('manifest/js.json')
	;

	return gulp.src(['**/*.php', '**/*.css'], {cwd: CONFIG.output})
	.pipe( plugins.revReplace({
		replaceInExtensions: ['.php', '.css'],
		manifest: manifestCss
	}) )
	.pipe( plugins.revReplace({
		replaceInExtensions: ['.php', '.css'],
		manifest: manifestImages
	}) )
	.pipe( plugins.revReplace({
		replaceInExtensions: ['.php'],
		manifest: manifestJs
	}) )
	.pipe( plugins.debug({title: 'revReplace + manifest'}) )
	.pipe( gulp.dest(CONFIG.output) )
	;
});


/* Задача для слежения за измениями в исходных файлах */
gulp.task('watch', function() {
	/* Копирование, когда изменились картинки  */
	gulp.watch('**/img/*.{jpg,png,gif,svg}', {cwd: CONFIG.blocks}, ['images']);

	/* Пересборка CSS, когда изменились стили  */
	gulp.watch('**/*.scss', {cwd: CONFIG.input}, ['styles:css']);

	/* Пересборка JS, когда изменились скрипты  */
	gulp.watch('**/*.js', {cwd: CONFIG.input}, ['scripts']);

	/* Пересборка HTML, когда изменились страницы, шаблоны или блоки */
	gulp.watch('**/*.php', {cwd: CONFIG.input}, ['nunjucks', 'revreplace']);

	/* Обработка остальных файлов */
	gulp.watch('**/*.*', {cwd: CONFIG.pages}, ['other.pages']);

	if ( CONFIG.reload ) {
		/* Перезагрузка браузера, когда что-то изменилось в сборке */
		gulp.watch(['**/*.*', '!.distr/**'], {cwd: CONFIG.output}).on('change', browserSync.reload);
	}
});


/* Задача для конечной сборки (для prod) */
gulp.task('build', function(){
	runSequence.options.ignoreUndefinedTasks = true;

	return runSequence(
		'clean',
		'images',
		['styles:css', 'scripts', 'other.pages'],
		'nunjucks',
		!isDevelopment ? 'revreplace' : null
		);
});


/* Задача по умолчанию (для dev) */
gulp.task('default', function(){
	return runSequence(
		'build',
		'browserSync',
		'watch'
		);
});


/* Задача для сборки без поднятия сервера */
gulp.task('nosync', function(){
	return runSequence(
		'build',
		'watch'
		);
});

/* Задача для сборки без перезагрузки */
gulp.task('noreload', function(){
	CONFIG.reload = false;

	return runSequence(
		'default'
		);
});
